src/main/scala/org/broadinstitute/hail/asm4s/FunctionBuilder.scala

```diff
@@ -88,110 +88,402 @@ abstract class FunctionBuilder[R](parameterTypeInfo: Array[TypeInfo[_]], returnT
+      //     out.write(sw.toString())
+      //     out.close()
+      //   } catch {
+      //     case e: IOException => System.out.println("Exception " + e)
+      //   }
+      //   println(sw)
+      //   println("Bytecode failed verification 1")
+      // }
+    }
+    {
+      val cw = new ClassWriter(ClassWriter.COMPUTE_MAXS + ClassWriter.COMPUTE_FRAMES)
+      cn.accept(cw)
+
+      val sw = new StringWriter()
+      CheckClassAdapter.verify(new ClassReader(cw.toByteArray()), false, new PrintWriter(sw))
+      if (sw.toString().length() != 0) {
+        println("Verify Output for " + name + ":")
+        try {
+          val out = new BufferedWriter(new FileWriter("ByteCodeOutput.txt"))
+          out.write(sw.toString())
+          out.close()
+        } catch {
+          case e: IOException => System.out.println("Exception " + e)
+        }
+        println(sw)
+        throw new IllegalStateException("Bytecode failed verification 2")
+      }
+    }
+
     val cw = new ClassWriter(ClassWriter.COMPUTE_MAXS +
       ClassWriter.COMPUTE_FRAMES)
     cn.accept(cw)
     val b = cw.toByteArray
-    val clazz = loadClass(null, b)
-
-    // print bytecode for debugging
-    /*
-    val cr = new ClassReader(b)
-    val tcv = new TraceClassVisitor(null, new Textifier, new PrintWriter(System.out))
-    cr.accept(tcv, 0)
-    */
-
-    clazz
-  }
-
-  def whileLoop(condition: Code[Boolean], body: Code[_]): Code[Unit] = {
-    val l1 = new LabelNode
-    val l2 = new LabelNode
-    new Code[Unit] {
-      def emit(il: InsnList): Unit = {
-        il.add(l1)
-        condition.emit(il)
-        il.add(new LdcInsnNode(0))
-        il.add(new JumpInsnNode(IF_ICMPEQ, l2))
-        body.emit(il)
-        il.add(new JumpInsnNode(GOTO, l1))
-        il.add(l2)
+    // val clazz = loadClass(null, b)
+
+    // clazz
+    b
+  }
+}
+
+class Function0Builder[R >: Null](implicit rti: TypeInfo[R]) extends FunctionBuilder[R](Array[TypeInfo[_]](), rti) {
+
+  cn.interfaces.asInstanceOf[java.util.List[String]].add("scala/Function0")
+
+  def result(c: Code[R]): () => R = {
+    val bytes = classAsBytes(c)
+    val localName = name.replaceAll("/",".")
+
+    new Function0[R] with java.io.Serializable {
+      @transient @volatile private var f: () => Any = null
+      def apply(): R = {
+        if (f == null) {
+          this.synchronized {
+            if (f == null) {
+              f = loadClass(localName, bytes).newInstance().asInstanceOf[() => Any]
+            }
+          }
+        }
+
+        f().asInstanceOf[R]
       }
     }
   }
 }
 
-class Function0Builder[R](implicit rti: TypeInfo[R]) extends FunctionBuilder[R](Array[TypeInfo[_]](), rti) {
-  def result(c: Code[R]): () => R = {
-    val clazz = resultClass(c)
-    val m = clazz.getMethod("f")
-    () => m.invoke(null).asInstanceOf[R]
+trait ToI { def apply(): Int }
+class FunctionToIBuilder(implicit rti: TypeInfo[Int]) extends FunctionBuilder[Int](Array[TypeInfo[_]](), rti) {
```

Comment:
I had to specialize for primitives to get the interfaces to work out correctly. In particular, the `Function1[T, U]`, as byte code, always takes a reference type. Which means I'd need a way to box primitives before passing them to invocations of `f`. Not sure how to do that, so i just specialized for every type.